pkg/utils/auth.go

```diff
@@ -406,3 +419,104 @@ func getCredentialAK() AccessControl {
 	config := sdk.NewConfig().WithScheme(scheme)
 	return AccessControl{Config: config, Credential: credential, UseMode: Credential}
 }
+
+type managedAddonTokenCredv2 struct {
+	sync.Mutex
+	token        *ManageTokens
+	lastUpdateAt time.Time
+	scale        float64
+}
+
+func newManagedAddonTokenCredv2() *managedAddonTokenCredv2 {
+	return &managedAddonTokenCredv2{
+		scale: addonTokenExpirationScale,
+	}
+}
+
+func (cred *managedAddonTokenCredv2) needUpdate() bool {
+	if cred.token == nil {
+		return true
+	}
+	duration := time.Since(cred.lastUpdateAt)
+	expiration := cred.token.ExpireAt.Sub(cred.lastUpdateAt)
+	return duration >= time.Duration(float64(expiration)*cred.scale)
```

Comment:
I think my version at https://github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pull/873/files#diff-7292cc76b3b0f97dd0d6676f1d72baa164adebc25b265417dd5dfc724236e780R152-R159 is better.

Since this is a local file, reloading has relatively low cost, so I just reload every 5 minutes, in case of user re-created the RAM role so that the token is invalidated earlier than expected. And I add a fixed 10 minutes offset to the expiration time, which I think is more reasonable than scale. Because this offset is used to compensate for clock drift, and this drift should not relate to the token valid duration.

Or are you willing to use my version of v2 cred in #873 ? I (or you) can cherry-pick the relevant commits.

Reply:
Good! I will take a look.